FSRS-5

joshdavham opened the pull request.

joshdavham commented:
@L-M-Sherlock Hey Jarrett, I was wondering if I could get some help with this.

I spent some time looking at the Rust and Typescript implementations of FSRS-5 and am feeling a bit confused about how to implement the new short-term stability function.

Also, is the new argument passed to the `mean_reversion` function correct? I couldn't find any note about changes on the [FSRS-5](https://github.com/open-spaced-repetition/fsrs4anki/wiki/The-Algorithm#fsrs-5-draft) page, but did find these two PRs that made changes: [ts-fsrs](https://github.com/open-spaced-repetition/ts-fsrs/commit/231539e9ce20994bcc68e2ccede0224ff1de792f), [fsrs-rs](https://github.com/open-spaced-repetition/fsrs-rs/pull/205).

L-M-Sherlock commented:
Sorry, I forget to update the wiki page. Now it shows the latest formula about mean reversion.

> feeling a bit confused about how to implement the new short-term stability function

The new short-term stability function is executed in learning stage:

https://github.com/open-spaced-repetition/ts-fsrs/blob/eee4f0c84f91d36f06f8fd55830f319d6a98e146/src/fsrs/impl/basic_schduler.ts#L69

joshdavham commented:
@L-M-Sherlock Before I continue with this PR, could I get you to take a quick look at whether the new mean reversion and short term stability formulas are applied correctly?

ishiko732 commented:
> @L-M-Sherlock Before I continue with this PR, could I get you to take a quick look at whether the new mean reversion and short term stability formulas are applied correctly?

I think you can refer to https://github.com/open-spaced-repetition/ts-fsrs/pull/95/files and https://github.com/open-spaced-repetition/ts-fsrs/pull/103/files to implement FSRS-v5.

joshdavham commented:
@ishiko732 

I'm currently testing this implementation using the same weights and ratings that you were using [here](https://github.com/open-spaced-repetition/ts-fsrs/blob/65fd676414e23e21612b5344af947480f7dafa7e/__tests__/FSRSV5.test.ts#L12-L37) and I'm getting different results.

While you were getting these resulting intervals:
```
[0, 4, 17, 62, 198, 563, 0, 0, 9, 27, 74, 190, 457]
```
I'm currently getting:
```
[0, 4, 16, 58, 183, 516, 0, 0, 17, 50, 144, 380, 928]
```

I went and compared the pre-FSRS-5 implementations of both py-fsrs and ts-fsrs (py-fsrs 2.5.1 and ts-fsrs 3.5.7) and verified that they produce identical intervals so this means that there's currently either something wrong with this fsrs-5 implementation or your ts-fsrs implementation... and I'm not sure which!

---
Furthermore, I tested these weights and ratings again, but with `w17=0 and w18=0` to 'disable' the new [short term stability function](https://github.com/open-spaced-repetition/fsrs4anki/wiki/The-Algorithm#formula) and still got different results:

ts-fsrs:
```
[0, 3, 13, 50, 163, 473, 0, 0, 12, 34, 91, 229, 541]
```
py-fsrs:
```
[0, 3, 13, 48, 155, 445, 0, 0, 12, 39, 116, 313, 780]
```
So this likely means that the cause of the problem isn't the short term stability problem alone.

Finally, I played around with ts-fsrs's short term and long term schedulers and I didn't end up making any more progress finding the issue :(

---
Any idea what might be causing the discrepency between our implementations?

ishiko732 commented on src/fsrs/fsrs.py:
```diff
@@ -58,9 +58,18 @@ def repeat(
             s.easy.scheduled_days = easy_interval
             s.easy.due = now + timedelta(days=easy_interval)
         elif card.state == State.Learning or card.state == State.Relearning:
+            short_term_stability_good = self.short_term_stability(
+                card.stability, Rating.Good
+            )
+            short_term_stability_easy = self.short_term_stability(
+                card.stability, Rating.Easy
+            )
+
```
Could it be due to not updating the difficulty? See here for more details: https://github.com/open-spaced-repetition/ts-fsrs/blob/65fd676414e23e21612b5344af947480f7dafa7e/src/fsrs/impl/basic_schduler.ts#L68-L111.

joshdavham replied:
I went and added
```python
  s.again.difficulty = self.next_difficulty(card.difficulty, Rating.Again)
  s.hard.difficulty = self.next_difficulty(card.difficulty, Rating.Hard)
  s.good.difficulty = self.next_difficulty(card.difficulty, Rating.Good)
  s.easy.difficulty = self.next_difficulty(card.difficulty, Rating.Easy)
```
just below and now when I set `w17=0 and w18=0`, we get the same results. So I think this means that the rest of the issue is left in one of our short term stability functions (probably mine)

joshdavham replied:
Okay I think I got it! I'll send some commits in a bit

joshdavham commented:
Alrighty I think this PR is ready for review!

To summarize this PR:
- Modified the default weights to the new FSRS-v5 [weights](https://github.com/open-spaced-repetition/fsrs4anki/wiki/The-Algorithm#default-parameters)
- Implemented the new mean reversion, initial difficulty, short term stability, formulas and now update the difficulties of cards in the Learning and Relearning states
- Updated the README custom weight example to use 19 parameters instead of 17
- Modified tests and made sure that ts-fsrs and py-fsrs give the same results
  - I could probably add some more tests to demonstrate that ts-fsrs (the [ground truth](https://github.com/open-spaced-repetition/ts-fsrs/issues/115#issuecomment-2295576965)) and py-fsrs give the same results, but I've also independently tested both implementations with different weight combinations and they always produce the same intervals
- Bump major version
  - 2.5.1 -> 3.0.0

I'd like some feedback on whether a major version bump here seems appropriate. This PR should only introduce breaking changes for those who were using custom weights, but this is likely very few people in reality. But on the other hand, these are still breaking changes and also it looks like both [ts-fsrs](https://github.com/open-spaced-repetition/ts-fsrs/releases/tag/v4.0.0-beta) and [fsrs-rs](https://github.com/open-spaced-repetition/fsrs-rs/releases/tag/v1.0.0) did major version bumps when introducing fsrs-5.

I'd also like to add a release note for those who may have previously been using custom weights. What do you suggest I write? I was thinking of suggesting setting w17 and w18=0 if they haven't yet computed their new fsrs-5 parameters. I could also point them to the [fsrs-optimizer](https://github.com/open-spaced-repetition/fsrs-optimizer) or suggest they revert back to the default parameters for now?

ishiko732 commented on README.md:
```diff
@@ -88,23 +88,25 @@ You can initialize the FSRS scheduler with your own custom weights as well as de
-        0.3976,
-        0.0,
-        2.0902,
+        0.4197,
+        1.1869,
+        3.0412,
+        15.2441,
+        7.1434,
+        0.6477,
+        1.0007,
+        0.0674,
+        1.6597,
+        0.1712,
+        1.1178,
+        2.0225,
+        0.0904,
+        0.3025,
+        2.1214,
+        0.2498,
+        2.9466,
+        0.4891,
+        0.6468,
```
I believe the default parameters should be consistent with the documentation.
- https://github.com/open-spaced-repetition/fsrs4anki/wiki/The-Algorithm#default-parameters
- https://github.com/open-spaced-repetition/fsrs-rs/blob/840d80e8b017e78af0f7484f11d500adc5a06388/src/inference.rs#L24-L27

joshdavham replied:
Yeah that works too. I initially was thinking of specifying different parameters in the README example just to show clearly that you can specify parameters different than the defaults, but this is fine 👍

ishiko732 commented:
> I was thinking of suggesting setting w17 and w18=0 if they haven't yet computed their new fsrs-5 parameters.

In this case, both ts-fsrs and fsrs-rs default to adding two [0.0,0.0].
- https://github.com/open-spaced-repetition/ts-fsrs/blob/65fd676414e23e21612b5344af947480f7dafa7e/src/fsrs/default.ts#L16-L35
- https://github.com/open-spaced-repetition/fsrs-rs/blob/840d80e8b017e78af0f7484f11d500adc5a06388/src/model.rs#L242-L248

joshdavham commented:
That's a cool idea, but I'm not sure if I want to autofill the last two parameters with 0's if the user specifies 17 parameters instead of 19. I think this could cause some unintended behavior for the user, so I'd personally prefer not to implement that in py-fsrs.

ishiko732 commented:
@L-M-Sherlock cc

L-M-Sherlock commented:
> whether a major version bump here seems appropriate.

It's appropriate. 

> I'd also like to add a release note for those who may have previously been using custom weights. What do you suggest I write?

We need to inform them that only v5 optimizer could generate correct parameters for them.

joshdavham commented:
> We need to inform them that only v5 optimizer could generate correct parameters for them.

For sure. I could mention to use the latest version of either ts-fsrs or fsrs-rs

ishiko732 commented:
> I could mention to use the latest version of either [fsrs-optimizer](https://github.com/open-spaced-repetition/fsrs-optimizer) or [fsrs-rs](https://github.com/open-spaced-repetition/fsrs-rs) for the optimizer

By the way: [fsrs-browser](https://github.com/open-spaced-repetition/fsrs-browser) can also be optimized.😊

ishiko732 reviewed (approved):
LGTM
